server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -99,9 +99,14 @@ public DesiredBalanceShardsAllocator(
         this.desiredBalanceComputer = desiredBalanceComputer;
         this.desiredBalanceComputation = new ContinuousComputation<>(threadPool.generic()) {
 
+            private long lastProcessedIndex = -1;
```

Comment:
I believe this should be resettled in case node is no longer elected master.
Otherwise below assertion would fail if the node is elected as master again in some point in future.

Alternatively we might reconsider resetting the index in `org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceShardsAllocator#onNoLongerMaster`

Reply:
Ah good point.

This seems tricky to make concurrency-safe, however - we might stop being master just before `processInput` is called for the last time. I'll think about this some more.

Reply:
Do we actually need to reset the `indexGenerator` when standing down as master? I didn't notice we did this, and it seems strange: we don't guarantee that everything gets properly cleared out in this case, so e.g. `ReconcileDesiredBalanceExecutor#findLatest` might pick a task from the previous master term.